File: framework/languages/java/zlink-stream-connector/src/main/java/systems/zlink/stream/connector/ZLinkStreamConnectorOptions.java
package systems.zlink.stream.connector;

import java.net.URI;
import java.time.Duration;

public record ZLinkStreamConnectorOptions(
        URI endpoint,
        ZLinkStreamDispatchMode dispatchMode,
        Duration requestTimeout,
        Duration waitTimeout,
        int maxReconnectAttempts,
        Duration connectTimeout,
        int maxSendPayloadSize,
        int maxReceivePayloadSize,
        boolean heartbeatEnabled,
        Duration heartbeatInterval,
        Duration heartbeatTimeout,
        boolean reconnectEnabled,
        Duration reconnectInitialDelay,
        Duration reconnectMaxDelay,
        double reconnectBackoffFactor,
        boolean skipServerCertificateValidation,
        ZLinkStreamCompression compression,
        ZLinkStreamCompressionCodec compressionCodec,
        ZLinkStreamPacketNameResolver nameResolver,
        ZLinkStreamTypedCodec typedCodec) {
    public static final int UNLIMITED_RECONNECT_ATTEMPTS = -1;

    public static ZLinkStreamConnectorOptions createDefault(URI endpoint) {
        return new ZLinkStreamConnectorOptions(
                endpoint,
                ZLinkStreamDispatchMode.MANUAL,
                Duration.ofSeconds(30),
                Duration.ofSeconds(5),
                3,
                Duration.ofSeconds(5),
                64 * 1024,
                64 * 1024,
                true,
                Duration.ofSeconds(1),
                Duration.ofSeconds(5),
                true,
                Duration.ofMillis(250),
                Duration.ofSeconds(5),
                2.0,
                false,
                ZLinkStreamCompression.LZ4,
                null,
                ZLinkStreamPacketNameResolver.defaultResolver(),
                null);
    }

    public ZLinkStreamConnectorOptions {
        if (compression == null) {
            compression = ZLinkStreamCompression.LZ4;
        }
        if (compression == ZLinkStreamCompression.NONE && compressionCodec != null) {
            //  Common connector spec §6.3 names this as a disagreement
            //  between two options, so the code is CONFIGURATION_ERROR and
            //  §9.2 requires it to travel in an exception that carries it.
            throw ZLinkStreamException.configurationError(
                    "compressionCodec cannot be set when compression is none");
        }
        if (compression == ZLinkStreamCompression.LZ4 && compressionCodec == null) {
            compressionCodec = ZLinkStreamCompressionCodecs.lz4();
        }
        if (nameResolver == null) {
            nameResolver = ZLinkStreamPacketNameResolver.defaultResolver();
        }
        if (typedCodec == null) {
            typedCodec = ZLinkStreamJson.codec();
        }
    }

    public ZLinkStreamConnectorOptions(
            URI endpoint,
            ZLinkStreamDispatchMode dispatchMode,
            Duration requestTimeout,
            int maxReconnectAttempts) {
        this(
                endpoint,
                dispatchMode,
                requestTimeout,
                Duration.ofSeconds(5),
                maxReconnectAttempts,
                Duration.ofSeconds(5),
                64 * 1024,
                64 * 1024,
                true,
                Duration.ofSeconds(1),
                Duration.ofSeconds(5),
                true,
                Duration.ofMillis(250),
                Duration.ofSeconds(5),
                2.0,
                false,
                ZLinkStreamCompression.LZ4,
                null,
                ZLinkStreamPacketNameResolver.defaultResolver(),
                null);
    }

    public ZLinkStreamConnectorOptions(
            URI endpoint,
            ZLinkStreamDispatchMode dispatchMode,
            Duration requestTimeout,
            int maxReconnectAttempts,
            Duration connectTimeout,
            int maxSendPayloadSize,
            boolean heartbeatEnabled,
            Duration heartbeatInterval,
            Duration heartbeatTimeout,
            boolean reconnectEnabled,
            Duration reconnectInitialDelay,
            Duration reconnectMaxDelay,
            double reconnectBackoffFactor) {
        this(
                endpoint,
                dispatchMode,
                requestTimeout,
                Duration.ofSeconds(5),
                maxReconnectAttempts,
                connectTimeout,
                maxSendPayloadSize,
                64 * 1024,
                heartbeatEnabled,
                heartbeatInterval,
                heartbeatTimeout,
                reconnectEnabled,
                reconnectInitialDelay,
                reconnectMaxDelay,
                reconnectBackoffFactor,
                false,
                ZLinkStreamCompression.LZ4,
                null,
                ZLinkStreamPacketNameResolver.defaultResolver(),
                null);
    }

    public ZLinkStreamConnectorOptions(
            URI endpoint,
            ZLinkStreamDispatchMode dispatchMode,
            Duration requestTimeout,
            int maxReconnectAttempts,
            Duration connectTimeout,
            int maxSendPayloadSize,
            boolean heartbeatEnabled,
            Duration heartbeatInterval,
            Duration heartbeatTimeout,
            boolean reconnectEnabled,
            Duration reconnectInitialDelay,
            Duration reconnectMaxDelay,
            double reconnectBackoffFactor,
            boolean skipServerCertificateValidation) {
        this(
                endpoint,
                dispatchMode,
                requestTimeout,
                Duration.ofSeconds(5),
                maxReconnectAttempts,
                connectTimeout,
                maxSendPayloadSize,
                64 * 1024,
                heartbeatEnabled,
                heartbeatInterval,
                heartbeatTimeout,
                reconnectEnabled,
                reconnectInitialDelay,
                reconnectMaxDelay,
                reconnectBackoffFactor,
                skipServerCertificateValidation,
                ZLinkStreamCompression.LZ4,
                null,
                ZLinkStreamPacketNameResolver.defaultResolver(),
                null);
    }

    public ZLinkStreamConnectorOptions(
            URI endpoint,
            ZLinkStreamDispatchMode dispatchMode,
            Duration requestTimeout,
            int maxReconnectAttempts,
            Duration connectTimeout,
            int maxSendPayloadSize,
            boolean heartbeatEnabled,
            Duration heartbeatInterval,
            Duration heartbeatTimeout,
            boolean reconnectEnabled,
            Duration reconnectInitialDelay,
            Duration reconnectMaxDelay,
            double reconnectBackoffFactor,
            ZLinkStreamTypedCodec typedCodec) {
        this(
                endpoint,
                dispatchMode,
                requestTimeout,
                Duration.ofSeconds(5),
                maxReconnectAttempts,
                connectTimeout,
                maxSendPayloadSize,
                64 * 1024,
                heartbeatEnabled,
                heartbeatInterval,
                heartbeatTimeout,
                reconnectEnabled,
                reconnectInitialDelay,
                reconnectMaxDelay,
                reconnectBackoffFactor,
                false,
                ZLinkStreamCompression.LZ4,
                null,
                ZLinkStreamPacketNameResolver.defaultResolver(),
                typedCodec);
    }

    public ZLinkStreamConnectorOptions(
            URI endpoint,
            ZLinkStreamDispatchMode dispatchMode,
            Duration requestTimeout,
            int maxReconnectAttempts,
            Duration connectTimeout,
            int maxSendPayloadSize,
            boolean heartbeatEnabled,
            Duration heartbeatInterval,
            Duration heartbeatTimeout,
            boolean reconnectEnabled,
            Duration reconnectInitialDelay,
            Duration reconnectMaxDelay,
            double reconnectBackoffFactor,
            boolean skipServerCertificateValidation,
            ZLinkStreamPacketNameResolver nameResolver) {
        this(
                endpoint,
                dispatchMode,
                requestTimeout,
                Duration.ofSeconds(5),
                maxReconnectAttempts,
                connectTimeout,
                maxSendPayloadSize,
                64 * 1024,
                heartbeatEnabled,
                heartbeatInterval,
                heartbeatTimeout,
                reconnectEnabled,
                reconnectInitialDelay,
                reconnectMaxDelay,
                reconnectBackoffFactor,
                skipServerCertificateValidation,
                ZLinkStreamCompression.LZ4,
                null,
                nameResolver,
                null);
    }
}

File: framework/languages/unity/com.zlink.stream-connector.webgl/Runtime/Contracts/ZlinkStreamConnectorOptions.cs
using System;

namespace Systems.Zlink.Stream.Connector.Contracts
{
    public sealed class ZlinkStreamConnectorOptions
    {
        /// <summary>
        ///     The STREAM endpoint. WebGL accepts <c>ws://</c> and <c>wss://</c> only;
        ///     <c>tcp://</c> and <c>tls://</c> fail with
        ///     <see cref="ZlinkStreamErrorCode.ConfigurationError" /> when the connector is
        ///     created, not later at runtime (stream-connector spec 32 section 3.2).
        /// </summary>
        /// <remarks>
        ///     The native .NET package marks this member <c>required</c>. Unity's C# 9
        ///     compiler has no <c>required</c> modifier, so the connector factory rejects a
        ///     missing endpoint instead.
        /// </remarks>
        public Uri Endpoint { get; init; }

        public ZlinkStreamTransport? Transport { get; init; }

        public TimeSpan ConnectTimeout { get; init; } = TimeSpan.FromSeconds(5);

        public TimeSpan RequestTimeout { get; init; } = TimeSpan.FromSeconds(30);

        public TimeSpan WaitTimeout { get; init; } = TimeSpan.FromSeconds(5);

        public ZlinkStreamHeartbeatOptions Heartbeat { get; init; } =
            new ZlinkStreamHeartbeatOptions();

        public ZlinkStreamReconnectOptions Reconnect { get; init; } =
            new ZlinkStreamReconnectOptions();

        public int MaxSendPayloadSize { get; init; } = 64 * 1024;

        public int MaxReceivePayloadSize { get; init; } = 64 * 1024;

        /// <summary>
        ///     Ignored on WebGL. The browser owns TLS verification for <c>wss://</c> and
        ///     offers no way to skip it, so setting this to <c>true</c> is rejected as a
        ///     configuration error instead of being silently ignored.
        /// </summary>
        public bool SkipServerCertificateValidation { get; init; }

        public ZlinkStreamDispatchMode DispatchMode { get; init; } = ZlinkStreamDispatchMode.Manual;

        public ZlinkStreamCompression Compression { get; init; } = ZlinkStreamCompression.Lz4;

        /// <summary>
        ///     Not supported on WebGL: the compression codec lives in the JavaScript
        ///     connector and a managed codec cannot cross the jslib boundary. Select the
        ///     algorithm with <see cref="Compression" /> instead. Setting this member makes
        ///     connector creation fail with
        ///     <see cref="ZlinkStreamErrorCode.ConfigurationError" />.
        /// </summary>
        public IZlinkStreamCompressionCodec CompressionCodec { get; init; }

        public IZlinkStreamPacketNameResolver NameResolver { get; init; } =
            new ZlinkStreamPacketNameResolver();

        /// <summary>
        ///     Payload codec for the typed connector API. Unlike the native package there is
        ///     no JSON fallback (Unity ships no System.Text.Json), so the typed API throws
        ///     <see cref="NotSupportedException" /> while this is unset.
        /// </summary>
        public IZlinkStreamPayloadCodec PayloadCodec { get; init; }
    }

    public sealed class ZlinkStreamHeartbeatOptions
    {
        public bool Enabled { get; init; } = true;

        public TimeSpan Interval { get; init; } = TimeSpan.FromSeconds(1);

        public TimeSpan Timeout { get; init; } = TimeSpan.FromSeconds(5);
    }

    public sealed class ZlinkStreamReconnectOptions
    {
        public bool Enabled { get; init; } = true;

        public TimeSpan InitialDelay { get; init; } = TimeSpan.FromMilliseconds(250);

        public TimeSpan MaxDelay { get; init; } = TimeSpan.FromSeconds(5);

        public double BackoffFactor { get; init; } = 2.0;

        public int? MaxAttempts { get; init; } = 3;
    }
}
